cache/pmdk_cache.cc

```diff
@@ -0,0 +1,438 @@
+    });
+  } else {
+    pop_ = po::pool<PersistentRoot>::open(PHEAP_PATH, "pmdk_cache_pool");
+    persistent_hashtable_ = new PersistTierHashTable(pop_.root()->persistent_hashtable);
+    lru_ = pop_.root()->persistent_lru_list;
+  }
+
+}
+
+void PMDKCacheShard::EraseUnRefEntries() {
+  // TODO
+}
+
+void PMDKCacheShard::ApplyToAllCacheEntries(void (*callback)(void*, size_t),
+                                           bool thread_safe) {
+  // TODO
+}
+
+void PMDKCacheShard::TEST_GetLRUList(TransientHandle** lru, TransientHandle** lru_low_pri) {
+  // MutexLock l(&mutex_);
+  // *lru = &lru_;
+  // *lru_low_pri = lru_low_pri_;
+  // TODO
+}
+
+size_t PMDKCacheShard::TEST_GetLRUSize() {
+  size_t lru_size = 0;
+  // TODO
+  return lru_size;
+}
+
+double PMDKCacheShard::GetHighPriPoolRatio() {
+  // TODO: return from transient tier.
+  MutexLock l(&mutex_);
+  return high_pri_pool_ratio_;
+}
+
+void PMDKCacheShard::LRU_Remove(po::persistent_ptr<PersistentEntry> e) {
+  assert(e->next_lru != nullptr);
+  assert(e->prev_lru != nullptr);
+  po::transaction::run(pop_, [&, e] {
+    e->next_lru->prev_lru = e->prev_lru;
+    e->prev_lru->next_lru = e->next_lru;
+    e->prev_lru = e->next_lru = nullptr;
+  });
+  size_t persist_charge = e->persist_charge;
+  assert(lru_usage_ >= persist_charge);
+  lru_usage_ -= persist_charge;
+}
+
+void PMDKCacheShard::LRU_Insert(po::persistent_ptr<PersistentEntry> e) {
+  assert(e->next_lru == nullptr);
+  assert(e->prev_lru == nullptr);
+  po::transaction::run(pop_, [&, e] {
+    // Inset "e" to head of LRU list.
+    e->next_lru = lru_;
+    e->prev_lru = lru_->prev_lru;
+    e->prev_lru->next_lru = e;
+    e->next_lru->prev_lru = e;
+  });
+  lru_usage_ += e->persist_charge;
+}
+
+void PMDKCacheShard::EvictFromLRU(size_t charge,
+                                 autovector<po::persistent_ptr<PersistentEntry>>* deleted) {
+  while ((usage_ + charge) > persistent_capacity_ && lru_->next_lru != lru_) {
+    po::persistent_ptr<PersistentEntry> old = lru_->next_lru;
+    // LRU list contains only elements which can be evicted
+    assert(old->InCache() && !old->HasRefs());
+    LRU_Remove(old);
+    persistent_hashtable_->Remove(old);
+    old->SetInCache(false);
+    size_t old_persist_charge = old->persist_charge;
+    assert(usage_ >= old_persist_charge);
+    usage_ -= old_persist_charge;
+    deleted->push_back(old);
+  }
+}
+
+TransientHandle* PMDKCacheShard::GetTransientHandle(po::persistent_ptr<PersistentEntry> e,
+                                            void* (*pack)(const Slice& slice)){
+  if (e == nullptr){
+    return nullptr;
+  }
+  TransientHandle* ret = e->trans_handle;
+  if (!ret){
+    // build a TransientHandle.
+    ret = new TransientHandle();
+    ret->key_data = e->key.get();
+    ret->key_length = e->key_size;
+    ret->value = pack(Slice(e->val.get(), e->val_size));
+    e->trans_handle = ret; // no need to be in transaction. It's transient.
+  }
+  return ret;
```

Comment:
We should check the `era`, right?

Reply:
true. I should put `era` check in the condition beside `!ret`